File: Parser.cpp
#include <string>
#include "chess_gui.h"

namespace Parser
{

/**
 * Helper methods for parse_algebraic_notation()
 */
    namespace
    {
        Move parse_normal_move(std::string notation);
        Move parse_destination_rank(std::string notation);
        Move parse_destination_file(std::string notation);
        Move parse_capture(std::string notation);
        Move parse_moving_piece(std::string notation);
        Move parse_source_rank(std::string notation);
        Move parse_source_file(std::string notation);
    }
/**
 * Parses a valid chess algebraic notation string into a Move struct.
 * 
 * @param notation a valid algebraic notation for a move
 * @return a Move that contains all the information represented in the notation
 * 
 * @throws TODO if notation is not valid
 */
Move parse_algebraic_notation(std::string notation)
{
    Move parsed_move;
    if (notation == "O-O" || notation == "0-0-0")
    {
        // castle king side
    }
    else if (notation == "O-O-O" || notation == "0-0-0")
    {
        // castle queen side
    }
    else if (notation == "ep" /* ends in e.p. */)
    {
        // en passant
    }
    else if (notation == "Q" /* ends in B, N, R, or Q */)
    {
        // promotion
    }
    else if (notation.back() == '+')
    {
        notation.pop_back();
        parsed_move = parse_normal_move(notation);
        parsed_move.move_type = parsed_move.move_type | MoveType::Check;
        return parsed_move;
    }
    else if (notation.back() == '#')
    {
        notation.pop_back();
        parsed_move = parse_normal_move(notation);
        parsed_move.move_type = parsed_move.move_type | MoveType::Checkmate;
        return parsed_move;
    }
    else
    {
        parsed_move = parse_normal_move(notation);
        return parsed_move; // TODO
    }
}

// TODO return reference to Move ?
// TODO use substring
namespace
{
/**
 * Parses algebraic notation that represents a non-castle, non-(en passant), non-promotion move.
 * 
 * @param notation a valid algebraic notation for a move
 * @return a Move that contains all the information represented in the notation
 */
Move parse_normal_move(std::string notation)
{
    if (notation.back() >= '1' && notation.back() <= '8')
    {
        Move parsed_move = parse_destination_rank(notation);
        parsed_move.move_type = parsed_move.move_type | MoveType::Normal;
        return parsed_move;
    }
    else
    {
        // throw
    }
}

/**
 * Parses algebraic notation whose last character represents the rank of the destination square.
 * 
 * @param notation a valid algebraic notation for a move
 * @return a Move that contains all the information represented in the notation
 */
Move parse_destination_rank(std::string notation)
{
    char destination_rank = notation.back() - '1';

    notation.pop_back();
    if (notation.back() >= 'a' && notation.back() <= 'h')
    {
        Move parsed_move = parse_destination_file(notation);
        parsed_move.destination.rank = (int) destination_rank;
        return parsed_move;
    }
    else
    {
        // throw
    }
}

/**
 * Parses algebraic notation whose last character represents the file of the destination square.
 * 
 * @param notation a valid partial algebraic notation for a move
 * @return a Move that contains all the information represented in the partial notation
 */
Move parse_destination_file(std::string notation)
{
    char destination_file = notation.back() - 'a';
    notation.pop_back();
    Move parsed_move;

    if (notation.empty() ||
        notation.back() == 'K' ||
        notation.back() == 'Q' ||
        notation.back() == 'R' ||
        notation.back() == 'B' ||
        notation.back() == 'N')
    {
        parsed_move = parse_moving_piece(notation);
    }
    else if (notation.back() == 'x' || notation.back() == 'X')
    {
        parsed_move = parse_capture(notation);
    }
    else if (notation.back() >= 'a' && notation.back() <= 'h')
    {
        parsed_move = parse_source_file(notation);
    }
    else if (notation.back() >= '1' && notation.back() <= '8')
    {
        parsed_move = parse_source_rank(notation);
    }
    else
    {
        // throw
    }

    parsed_move.destination.file = (int)destination_file;
    return parsed_move;
}

/**
 * Parses algebraic notation whose last character represents a capture.
 * 
 * @param notation a valid partial algebraic notation for a move
 * @return a Move that contains all the information represented in the partial notation
 */ 
Move parse_capture(std::string notation)
{
    notation.pop_back();
    Move parsed_move;
    if (notation.back() == 'K' ||
        notation.back() == 'Q' ||
        notation.back() == 'R' ||
        notation.back() == 'B' ||
        notation.back() == 'N')
    {

        parsed_move = parse_moving_piece(notation);
    }
    else if (notation.back() >= 'a' && notation.back() <= 'h')
    {
        parsed_move = parse_source_file(notation);
    }
    else if (notation.back() >= '1' && notation.back() <= '8')
    {
        parsed_move = parse_source_rank(notation);
    }
    else
    {
        // throw
    }

    parsed_move.move_type = parsed_move.move_type | MoveType::Capture;
    return parsed_move;
}

/**
 * Parses algebraic notation whose last character represents the rank of the source square.
 * 
 * @param notation a valid partial algebraic notation for a move
 * @return a Move that contains all the information represented in the partial notation
 */
Move parse_source_rank(std::string notation)
{
    char source_rank = notation.back() - '1';
    notation.pop_back();
    Move parsed_move;

    if (notation.back() == 'K' ||
        notation.back() == 'Q' ||
        notation.back() == 'R' ||
        notation.back() == 'B' ||
        notation.back() == 'N')
    {

        parsed_move = parse_moving_piece(notation);
    }
    else if (notation.back() >= 'a' && notation.back() <= 'h')
    {
        parsed_move = parse_source_file(notation);
    }
    else
    {
        // throw
    }

    parsed_move.source.rank = source_rank;
    return parsed_move;
}

/**
 * Parses algebraic notation whose last character represents the file of the source square.
 * 
 * @param notation a valid partial algebraic notation for a move
 * @return a Move that contains all the information represented in the partial notation
 */
Move parse_source_file(std::string notation)
{
    char source_file = notation.back() - 'a';
    notation.pop_back();
    Move parsed_move;

    if (notation.empty() ||
        notation.back() == 'K' ||
        notation.back() == 'Q' ||
        notation.back() == 'R' ||
        notation.back() == 'B' ||
        notation.back() == 'N')
    {

        parsed_move = parse_moving_piece(notation);
    }
    else
    {
        // throw
    }

    parsed_move.source.file = source_file;
    return parsed_move;
}

/**
 * Parses algebraic notation whose last character represents a chess piece.
 * 
 * @param notation a valid partial algebraic notation for a move
 * @return a Move that contains all the information represented in the partial notation
 */
Move parse_moving_piece(std::string notation)
{
    std::unordered_map<char, PieceType> piece_from_notation =
        {
            {'K', PieceType::King},
            {'Q', PieceType::Queen},
            {'B', PieceType::Bishop},
            {'N', PieceType::Knight},
            {'R', PieceType::Rook}};

    Move parsed_move{Coordinate{-1, -1}, Coordinate{-1, -1}, PieceType::None, MoveType::None};
    if (notation.empty())
    {
        parsed_move.piece_type = PieceType::Pawn;
    }
    else
    {
        char piece_notation = notation.back();
        parsed_move.piece_type = piece_from_notation[piece_notation];

        notation.pop_back();
        if (!notation.empty())
        {
            // throw
        }
    }

    return parsed_move;
}
}
}

File: src/com/keshane/Paw/Parser.java
package com.keshane.Paw;

import java.text.ParseException;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

class Parser {
    private static final String KINGSIDE_CASTLE_OHS = "O-O";
    private static final String KINGSIDE_CASTLE_ZEROES = "0-0";
    private static final String QUEENSIDE_CASTLE_OHS = "O-O-O";
    private static final String QUEENSIDE_CASTLE_ZEROES = "0-0-0";
    private static final String EN_PASSANT = "ep";

    private static final String[] pieceNotations = new String[]{Piece.Type.BISHOP.getNotation(),
            Piece.Type.KNIGHT.getNotation(), Piece.Type.ROOK.getNotation(), Piece.Type.QUEEN
            .getNotation(), Piece.Type.KING.getNotation()};
    private static final Set<String> PIECE_NOTATIONS = new HashSet<>(Arrays.asList
            (pieceNotations));
    private String notation;
    private Move.Builder move;

    Parser(String notation) throws ParseException {
        this.notation = notation;
        this.move = new Move.Builder();
        parseFullNotation(this.notation);
    }

    String getNotation() {
        return this.notation;
    }

    Move.Builder getMove() {
        return this.move;
    }

    // TODO consider using StringBuilder
    private void parseFullNotation(String notation) throws ParseException {

        // First, parse away unnecessary information
        if (notation.endsWith("+")) {
            move.addType(Move.Type.CHECK);
            notation = notation.substring(0, notation.length() - 1);
        }
        else if (notation.endsWith("#")) {
            move.addType(Move.Type.CHECKMATE);
            notation = notation.substring(0, notation.length() - 1);
        }

        if (notation.equalsIgnoreCase(Parser.KINGSIDE_CASTLE_OHS) || notation.equalsIgnoreCase(KINGSIDE_CASTLE_ZEROES)) {
            move.addType(Move.Type.KINGSIDE_CASTLE);
            return;
        }
        else if (notation.equalsIgnoreCase(Parser.QUEENSIDE_CASTLE_OHS) || notation.equalsIgnoreCase(Parser.QUEENSIDE_CASTLE_ZEROES)) {
            move.addType(Move.Type.QUEENSIDE_CASTLE);
            return;
        }
        else if (notation.endsWith(Parser.EN_PASSANT)) {
            move.addType(Move.Type.EN_PASSANT);
            notation = notation.substring(0, notation.length() - 1);
        }
        else if (notation.endsWith(Piece.Type.QUEEN.getNotation()) ||
                notation.endsWith(Piece.Type.BISHOP.getNotation()) ||
                notation.endsWith(Piece.Type.KNIGHT.getNotation()) ||
                notation.endsWith(Piece.Type.ROOK.getNotation())) {
            move.addType(Move.Type.PROMOTION);
            notation = notation.substring(0, notation.length() - 1);
        }
        else {
            move.addType(Move.Type.NORMAL);
        }

        if (notation.length() < 2) {
            throw new ParseException("Notation isn't long enough.", notation.length() - 1);
        }
        // start parsing the notation from the last character and moving toward the front
        parseDestinationRank(notation.substring(0, notation.length()));
    }

    /**
     * Parses the last character of {@code notation} as the rank of the destination square.
     *
     * @param notation the move notation whose last character is the destination rank
     */
    private void parseDestinationRank(String notation) throws ParseException {
        int destinationRank = notation.charAt(notation.length() - 1) - '1';

        if (destinationRank < 0 || destinationRank > 7) {
            throw new ParseException("Destination rank is not between '1' and '8'.", notation
                    .length() - 1);
        }

        move.setDestinationRank(destinationRank);
        parseDestinationFile(notation.substring(0, notation.length() - 1));
    }

    private void parseDestinationFile(String notation) throws ParseException {
        int destinationFile = notation.charAt(notation.length() - 1) - 'a';

        if (destinationFile < 0 || destinationFile > 7) {
            throw new ParseException("Destination file is not between 'a' and 'h'.", notation
                    .length() - 1);
        }

        move.setDestinationFile(destinationFile);
        parseLastSymbol(notation.substring(0, notation.length() - 1));
    }

    private void parseLastSymbol(String notation) throws ParseException {
        if (notation.length() == 0) {
            move.setPieceType(Piece.Type.PAWN);
            return;
        }
        char symbol = notation.charAt(notation.length() - 1);
        if ('x' == symbol) {
            parseCapture(notation);
        }
        else if (symbol >= 'a' && symbol <= 'h') {
            parseSourceFile(notation);
        }
        else if (symbol >= '1' && symbol <= '8') {
            parseSourceRank(notation);
        }
        else if (Parser.PIECE_NOTATIONS.contains(String.valueOf(symbol))) {
            parsePiece(notation);
        }
        else {
            throw new ParseException("Unrecognized symbol", notation.length() - 1);
        }
    }

    private void parseCapture(String notation) throws ParseException {
        move.addType(Move.Type.CAPTURE);

        parseLastSymbol(notation.substring(0, notation.length() - 1));
    }

    private void parseSourceFile(String notation) throws ParseException {
        int sourceFile = notation.charAt(notation.length() - 1) - 'a';

        if (sourceFile < 0 || sourceFile > 7) {
            throw new ParseException("Source file is not between 'a' and 'h'.", notation
                    .length() - 1);
        }
        move.setSourceFile(sourceFile);
        parseLastSymbol(notation.substring(0, notation.length() - 1));
    }

    private void parseSourceRank(String notation) throws ParseException {
        int sourceRank = notation.charAt(notation.length() - 1) - '1';

        if (sourceRank < 0 || sourceRank > 7) {
            throw new ParseException("Source rank is not between '1' and '8'.", notation
                    .length() - 1);
        }
        move.setSourceRank(sourceRank);
        parseLastSymbol(notation.substring(0, notation.length() - 1));
    }

    private void parsePiece(String notation) throws ParseException {
        char pieceNotation = notation.charAt(notation.length() - 1);

        move.setPieceType(Piece.Type.parse(String.valueOf(pieceNotation)));

        if (notation.length() != 1) {
            throw new ParseException("Unexpected additional symbols before piece notation.",
                    notation.length());
        }

    }


}



